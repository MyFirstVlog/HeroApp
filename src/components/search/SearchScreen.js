import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useNavigate, useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const {q = ''} = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({
        searchText: q,
    })

    const {searchText} = formValues;
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        console.log({searchText});
        navigate(`?q=${searchText}`)
    }
    return (
        <>
            <h1>Busqueda</h1>
            <hr />

            <div className='row'>                
                <div className='col-5'>                    
                    <h4>Custom Search</h4>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input 
                            type="text"
                            placeholder='Buscar un hÃ©roe'
                            className='form-control'
                            name= 'searchText'
                            autoComplete='off'
                            value={searchText}                            
                            onChange={handleInputChange}
                        />
                        <button
                            type='submit'
                            className='btn btn-outline-primary mt-1'
                        >
                            Buscar...
                        </button>
                    </form>                
                 </div>
                 <div className='col-7'>
                    <h4>Resultados</h4>

                    {
                        (q === '') ? <div className='alert alert-info'>
                            Buscar Un heroe
                        </div>
                        : (heroesFiltered.length === 0) && <div className='alert alert-danger'>No hay Resultados: {q}</div>
                    }

                    {
                       heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                        
                    }
                 </div>
           
           
           
           
           
           
            </div>
        </>
    )
}
