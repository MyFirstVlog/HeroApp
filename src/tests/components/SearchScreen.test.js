import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../components/search/SearchScreen"


const mockNavigate = jest.fn(); //Debe estar asi exactamente, si no, no sirve
//* Mocks
//? memory router
jest.mock('react-router-dom', ()=> ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))


describe('should test SearchScreen', () => {
    
    test('should show the component correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot(); 
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar Un heroe');
    });

    test('should show batman and input withn the value of query string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot(); 
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });

    test('should show error when thero is not found', () => {
        const queryParams = 'batman123';

        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${queryParams}`]}>
                <SearchScreen />
            </MemoryRouter>
        );
    
        expect(wrapper.find('.alert-danger')).toBeDefined();
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`No hay Resultados: ${queryParams}`)
    });

    test('should call navigate to the new url', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        ); 
        //? Es como si hubieramos escrito en el input batman
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman' 
            }
        })

        //? simular el submit del forms

        wrapper.find('form').prop('onSubmit')({
            preventDefault: ()=>{}
        })

        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
    })
    
    


    
    
})
