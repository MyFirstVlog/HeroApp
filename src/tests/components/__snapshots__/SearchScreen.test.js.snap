// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test SearchScreen should show batman and input withn the value of query string 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Custom Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <div
        className="col"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`should test SearchScreen should show the component correctly with default values 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Custom Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <div
        className="alert alert-info"
      >
        Buscar Un heroe
      </div>
    </div>
  </div>,
]
`;
